demystifying_regular_expressions/demystifying_regular_expressions.md

```diff
@@ -0,0 +1,305 @@
+
+</div>
+
+Using the pattern above, you could search for and find valid URLs (assuming we got the rules right) in your data. 
+
+In this module, we're not going to go into the rules for making a regular expression, and we won't be exhaustive in explaining each part of our complicated code. For now, here's a sneak preview of what each segment of this regular expression mean.  Don't worry about translating these or understanding them!
+
+* `http` : The literal characters "http"
+* `s?`: Optionally, the letter "s"
+* `:\/\/` : A colon and two forward slashes: "://"
+* `([A-Za-z0-9\-]+\.)?`: Optionally, a group of characters consisting of: 
+
+  - `[A-Za-z0-9\-]+` : One or more characters that's either in the A-Z range, the a-z range, the 0-9 range, or a literal hyphen (`-`)
+  - `\.` : A literal period (`.`)
+* `[A-Za-z0-9\-]+`: One or more characters that's either in the A-Z range, the a-z range, the 0-9 range, or a literal hyphen (`-`)
+* `\.` : A literal period (`.`)
+* `[A-Za-z0-9\-\.]+`: One or more characters that's either in the A-Z range, the a-z range, the 0-9 range, a literal hyphen (`-`), or a literal period (`.`)
+* `\/?`: Optionally, a final forward slash
+
+
+## A Great Tool to Start With
+
+As a final step in demystifying regular expressions, it can be helpful to introduce the online regex checker as a great resource to learn and to check your regex before trying to learn how to make your own regular expressions or use regex in SQL, R, or Python code you're writing.
+
+We like "Regex 101", at https://www.regex101.com, but you can also search online for "regex checker" to find other, similar websites.  Regex 101 allows you to enter your regular expression as well as some sample text that you want to apply the regular expression to.  
+
+Below, or by [trying it out yourself](https://regex101.com/r/QUok4D), you can see that we put in our "valid URL" regular expression (at the top of the screen, in the appropriately labeled box) as well as the different varieties of URLs, some of which we think are valid and match our pattern, and some of which we think are invalid and should fail to match our pattern.
+
+If our test strings match the pattern, they are highlighted, and if they do not match, they are not highlighted.
+
+![Regex 101 application with sample URL regex pattern and test URL strings.](media/url_regex.png)<!-- style = "border: 1px solid rgb(var(--color-highlight));" -->
+
+This tool is useful, because sometimes you may not need to construct a regular expression, because it was given to you in a code sample you found online, or in code someone else wrote, like a coworker in your lab.  You will still want to test that it works the way you expect, however, even if you don't totally understand what each character of the regular expression means.  
+
+## Quiz: Demystifying Regex
+
+
+Fill in the blank below with a single lowercase word.
+---
+
+Regular expressions are a way to define rules that describe a 
+
+[[ pattern ]] in text. 
```

Comment:
I feel like the this question needs a little restructuring. I'm not sure coming up with the word "pattern" is the most useful way to meet the learning objective of explaining what a regular expression is. I need to ponder what type of questions might get at this learning objective better and get back to you.

Reply:
Ok, I think I'd like this better as a "Check all that are true" question with options like:
- [[X]] Regular expressions are a way to define rules that describe a pattern in text.
- [[X]] Regular expressions can be used in multiple coding languages.
- [[X]] Regular expressions can be used to extract useful substrings.
- [[ ]]Regular expressions are easy to learn and simple to memorize.

Reply:
I agree!